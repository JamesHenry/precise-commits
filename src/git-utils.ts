import { sync as findUpSync } from 'find-up';
import { dirname } from 'path';

import { runCommandSync } from './utils';
import {
  hasPrettierSupportedFileExtension,
  generatePrettierIgnorePredicate,
} from './prettier';

interface DiffIndexFile {
  diffFilterChar: string;
  filename: string;
}

export function resolveNearestGitDirectory(workingDirectory: string) {
  return findUpSync('.git', { cwd: workingDirectory });
}

export function getDiffForFile(filename: string): string {
  return runCommandSync(
    'git',
    ['diff', '--unified=0', '--cached', filename],
    dirname(filename),
  ).stdout;
}

/**
 * NOTE: We are only explicitly testing "Modified" and "Added" files for now...
 * ----------------------------------------------------------------------
 *
 * Output of `git diff-index --help`:
 *
 * --diff-filter=[(A|C|D|M|R|T|U|X|B)...[*]]
      Select only files that are Added (A), Copied (C), Deleted (D), Modified (M), Renamed (R), have
      their type (i.e. regular file, symlink, submodule, ...) changed (T), are Unmerged (U), are
      Unknown (X), or have had their pairing Broken (B). Any combination of the filter characters
      (including none) can be used. When * (All-or-none) is added to the combination, all paths are
      selected if there is any file that matches other criteria in the comparison; if there is no
      file that matches other criteria, nothing is selected.

      Also, these upper-case letters can be downcased to exclude. E.g.  --diff-filter=ad excludes
      added and deleted paths.
 *
 */
const DIFF_INDEX_FILTER = 'ACDMRTUXB';
const SPECIAL_EMPTY_TREE_COMMIT_HASH =
  '4b825dc642cb6eb9a060e54bf8d69288fbee4904';

export function getRelevantModifiedFiles(
  workingDirectory: string,
  filesWhitelist: string[] | null,
  sha1?: string,
  sha2?: string,
): string[] {
  /**
   * Resolve the relevant .git directory
   */
  const gitDirectoryPath = resolveNearestGitDirectory(workingDirectory);
  if (!gitDirectoryPath) {
    throw new Error('No .git directory found');
  }
  const gitDirectoryParent = dirname(gitDirectoryPath);
  let diffIndexOutput: string;
  if (sha1 && sha2) {
    /**
     * We are grabbing the files between the two given commit SHAs
     */
    diffIndexOutput = runCommandSync(
      'git',
      [
        'diff',
        '--name-status',
        `--diff-filter=${DIFF_INDEX_FILTER}`,
        sha1,
        sha2,
      ],
      gitDirectoryParent,
    ).stdout;
  } else {
    /**
     * No commit SHAs given, we assume we are attempting to evaluate staged files,
     * and so need to determine if there is HEAD SHA available.
     */
    let head: string = '';
    try {
      head = runCommandSync(
        'git',
        ['rev-parse', '--verify', 'HEAD'],
        gitDirectoryParent,
      ).stdout.replace('\n', '');
    } catch (err) {
      /**
       * If there has never been a commit before, there will be no HEAD to compare
       * to. Use the special empty tree hash value instead:
       * https://stackoverflow.com/questions/9765453/is-gits-semi-secret-empty-tree-object-reliable-and-why-is-there-not-a-symbolic
       */
      if (err.message.includes(`fatal: Needed a single revision`)) {
        head = SPECIAL_EMPTY_TREE_COMMIT_HASH;
      } else {
        throw err;
      }
    }
    diffIndexOutput = runCommandSync(
      'git',
      [
        'diff-index',
        '--cached',
        '--name-status',
        `--diff-filter=${DIFF_INDEX_FILTER}`,
        head,
      ],
      gitDirectoryParent,
    ).stdout;
  }
  const allFiles = parseDiffIndexOutput(diffIndexOutput);
  /**
   * We fundamentally check whether or not the file extensions are supported by prettier,
   * whether or not they are included in the optional `filesWhitelist` array, and that the user
   * has not chosen to ignore them via a .prettierignore file.
   */
  return allFiles
    .map(r => r.filename)
    .filter(hasPrettierSupportedFileExtension)
    .filter(generateFilesWhitelistPredicate(filesWhitelist))
    .filter(generatePrettierIgnorePredicate(workingDirectory));
}

function generateFilesWhitelistPredicate(
  filesWhitelist: string[] | null,
): (file: string) => boolean {
  if (!filesWhitelist) {
    return () => true;
  }
  return file => !filesWhitelist.includes(file);
}

function parseDiffIndexOutput(stdout: string): DiffIndexFile[] {
  const lines = stdout.split('\n');
  return lines.filter(Boolean).map(line => {
    const parts = line.split('\t');
    return {
      filename: parts[1],
      diffFilterChar: parts[0],
    };
  });
}
