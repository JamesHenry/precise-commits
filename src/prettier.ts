import * as fs from 'fs';
import { extname, join } from 'path';
import {
  resolveConfig,
  format,
  check,
  getSupportInfo,
  Options,
} from 'prettier';

const ignore = require('ignore');
const DiffMatchPatch = require('diff-match-patch');
const dmp = new DiffMatchPatch();

/**
 * Resolve the user's prettier config
 */
export function resolvePrettierConfigForFile(filename: string): Options | null {
  return resolveConfig.sync(filename, { useCache: false });
}


/**
 * Run prettier on each character range pair given, and apply the
 * difference as a patch to the original contents using an implementation
 * of the Myer's diff algorithm.
 */
export function formatRangesWithinContents(
  characterRanges: any[],
  fileContents: string,
  prettierConfig: Options | null,
): string {
  let patches: any = [];
  characterRanges.forEach(characterRange => {
    const diffs = dmp.diff_main(
      fileContents,
      format(fileContents, {
        ...prettierConfig,
        ...{
          rangeStart: characterRange.rangeStart,
          rangeEnd: characterRange.rangeEnd,
        },
      }),
    );
    patches = [...patches, ...dmp.patch_make(fileContents, diffs)];
  });
  const [formattedContents] = dmp.patch_apply(patches, fileContents);
  return formattedContents;
}

/**
 * Ensure the given content has been run through prettier for the specific range data given.
 * This may require running prettier on the file multiple times for multiple ranges.
 */
export function checkRangesWithinContents(
  characterRanges: any[],
  fileContents: string,
  prettierConfig: Options | null,
): boolean {
  let formattedContents = fileContents;
  return characterRanges.every(characterRange => {
    return check(formattedContents, {
      ...prettierConfig,
      ...{
        rangeStart: characterRange.rangeStart,
        rangeEnd: characterRange.rangeEnd,
      },
    });
  });
}

/**
 * Return true if the full fileContents is already formatted according
 * to the given prettierConfig
 */
export function isAlreadyFormatted(
  fileContents: string,
  prettierConfig: Options | null,
): boolean {
  return check(fileContents, { ...prettierConfig });
}

let PRETTIER_SUPPORTED_FILE_EXTENSIONS: string[] = [];
getSupportInfo().languages.forEach(language => {
  PRETTIER_SUPPORTED_FILE_EXTENSIONS = [
    ...PRETTIER_SUPPORTED_FILE_EXTENSIONS,
    ...language.extensions,
  ];
});

export function hasPrettierSupportedFileExtension(filename: string) {
  const fileExtension = extname(filename);
  return PRETTIER_SUPPORTED_FILE_EXTENSIONS.includes(fileExtension);
}

export function generatePrettierIgnorePredicate(workingDirectory: string) {
  const prettierIgnoreFilePath = join(workingDirectory, '.prettierignore');
  /**
   * If there is no .prettierignore file present, simply always return true
   * from the predicate
   */
  if (!fs.existsSync(prettierIgnoreFilePath)) {
    return () => true;
  }
  /**
   * Use "ignore"'s createFilter() method to create a predicate
   */
  const prettierIgnoreFileContents = fs.readFileSync(
    prettierIgnoreFilePath,
    'utf8',
  );
  return ignore()
    .add(prettierIgnoreFileContents)
    .createFilter();
}
