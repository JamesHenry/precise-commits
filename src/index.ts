import { readFileSync, writeFileSync } from 'fs';
import { join } from 'path';

import { getDiffForFile, getRelevantModifiedFiles } from './git-utils';
import {
  extractLineChangeData,
  calculateCharacterRangesFromLineChanges,
  NO_LINE_CHANGE_DATA_ERROR,
  LineChangeData,
} from './utils';
import {
  formatRangesWithinContents,
  resolvePrettierConfigForFile,
  checkRangesWithinContents,
  isAlreadyFormatted,
} from './prettier';

export type ProcessingStatus = 'NOT_UPDATED' | 'UPDATED' | 'INVALID_FORMATTING';

export interface AdditionalOptions {
  checkOnly: boolean;
  filesWhitelist: string[] | null;
  sha1: string | null;
  sha2: string | null;
}

export interface Callbacks {
  onInit(workingDirectory: string): void;
  onModifiedFilesDetected(modifiedFilenames: string[]): void;
  onBegunProcessingFile(
    filename: string,
    index: number,
    totalFiles: number,
  ): void;
  onFinishedProcessingFile(
    filename: string,
    index: number,
    status: ProcessingStatus,
  ): void;
  onError(err: Error): void;
  onComplete(totalFiles: number): void;
}

function applyDefaults(options: AdditionalOptions): AdditionalOptions {
  options = options || {};
  options.filesWhitelist = options.filesWhitelist || null;
  options.sha1 = options.sha1 || null;
  options.sha2 = options.sha2 || null;
  options.checkOnly = !!options.checkOnly;
  return options;
}

/**
 * prettier-lines library
 */
export function main(
  workingDirectory: string,
  options: AdditionalOptions,
  callbacks: Callbacks = {
    onInit() {},
    onModifiedFilesDetected() {},
    onBegunProcessingFile() {},
    onFinishedProcessingFile() {},
    onError() {},
    onComplete() {},
  },
) {
  /**
   * Apply default options
   */
  const { checkOnly, filesWhitelist, sha1, sha2 } = applyDefaults(options);
  try {
    callbacks.onInit(workingDirectory);

    const modifiedFilenames = getRelevantModifiedFiles(
      workingDirectory,
      filesWhitelist,
      sha1,
      sha2,
    );
    callbacks.onModifiedFilesDetected(modifiedFilenames);
    const totalFiles = modifiedFilenames.length;

    modifiedFilenames.forEach((filename, index) => {
      callbacks.onBegunProcessingFile(filename, index, totalFiles);
      const fullPath = join(workingDirectory, filename);
      /**
       * Read the staged file contents and resolve the relevant prettier config
       */
      const fileContents = readFileSync(fullPath, 'utf8');
      const prettierConfig = resolvePrettierConfigForFile(fullPath);
      /**
       * To avoid unnecessary issues with 100% valid files producing issues when parts
       * of them are reformatted in isolation, we need to first check the whole file
       * to see if it is already formatted. This also allows us to skip unnecessary git
       * diff analysis work.
       */
      if (isAlreadyFormatted(fileContents, prettierConfig)) {
        return callbacks.onFinishedProcessingFile(
          filename,
          index,
          'NOT_UPDATED',
        );
      }
      /**
       * Extract line change data from the git diff results
       */
      const diff = getDiffForFile(fullPath, sha1, sha2);
      let lineChangeData: LineChangeData = { additions: [], removals: [] };
      try {
        lineChangeData = extractLineChangeData(diff);
      } catch (err) {
        if (err.message === NO_LINE_CHANGE_DATA_ERROR) {
          return callbacks.onFinishedProcessingFile(
            filename,
            index,
            'NOT_UPDATED',
          );
        }
        throw err;
      }
      /**
       * Convert the line change data into character data
       */
      const characterRanges = calculateCharacterRangesFromLineChanges(
        lineChangeData,
        fileContents,
      );
      /**
       * Run prettier on the file, multiple times if required, instructing it
       * to only focus on the calculated range(s).
       *
       * If `checkOnly` is set, only check to see if the formatting is valid,
       * otherwise automatically update the formatting.
       */
      /**
       * CHECK
       */
      if (checkOnly) {
        const isValid = checkRangesWithinContents(
          characterRanges,
          fileContents,
          prettierConfig,
        );
        if (!isValid) {
          return callbacks.onFinishedProcessingFile(
            filename,
            index,
            'INVALID_FORMATTING',
          );
        }
      }
      /**
       * FORMAT
       */
      const formattedFileContents = formatRangesWithinContents(
        characterRanges,
        fileContents,
        prettierConfig,
      );
      if (formattedFileContents === fileContents) {
        return callbacks.onFinishedProcessingFile(
          filename,
          index,
          'NOT_UPDATED',
        );
      }
      /**
       * Write the file back to disk
       */
      writeFileSync(fullPath, formattedFileContents);
      return callbacks.onFinishedProcessingFile(filename, index, 'UPDATED');
    });

    callbacks.onComplete(totalFiles);
  } catch (err) {
    callbacks.onError(err);
  }
}
