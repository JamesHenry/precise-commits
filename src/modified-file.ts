import { readFileSync, writeFileSync } from 'fs';

import {
  CharacterRange,
  extractLineChangeData,
  calculateCharacterRangesFromLineChanges,
} from './utils';
import { getDiffForFile } from './git-utils';
import { PreciseFormatter } from './precise-formatter';

export interface ModifiedFileConfig {
  fullPath: string;
  gitDirectoryParent: string;
  sha1: string | null;
  sha2: string | null;
  preciseFormatter: PreciseFormatter<any>;
}

export class ModifiedFile {
  private fullPath: string;
  /**
   * An optional commit SHA pair which will be used to inform how the git
   * commands are run. E.g. `git diff`
   */
  private sha1: string | null;
  private sha2: string | null;
  /**
   * The chosen formatter to be run on the modified file.
   */
  private preciseFormatter: PreciseFormatter<any>;
  /**
   * The parent directory of the relevant .git directory that was resolved
   * for the modified file.
   */
  private gitDirectoryParent: string;
  /**
   * The contents of the file in their current state on the user's file
   * system
   */
  public fileContents: string;
  /**
   * The final file contents, after we've run the formatter
   */
  private formattedFileContents: string;
  /**
   * The resolved formatter config which applies to this file
   */
  private formatterConfig: object | null;
  /**
   * The calculated character ranges which have been modified
   * within this file
   */
  private modifiedCharacterRanges: CharacterRange[] = [];

  constructor({
    fullPath,
    gitDirectoryParent,
    sha1,
    sha2,
    preciseFormatter,
  }: ModifiedFileConfig) {
    this.fullPath = fullPath;
    this.gitDirectoryParent = gitDirectoryParent;
    this.sha1 = sha1;
    this.sha2 = sha2;
    this.preciseFormatter = preciseFormatter;
    this.resolveFileContents();
    this.resolveFormatterConfig();
  }

  /**
   * Return true if the whole file has already been formatted appropriately based on
   * the resolved formatter config. We can use this as a check to skip unnecessary work.
   */
  isAlreadyFormatted(): boolean {
    return this.preciseFormatter.isAlreadyFormatted(
      this.fileContents,
      this.formatterConfig,
    );
  }

  /**
   * Run the formatters check mode on the given ranges and return true if they are all
   * already formatted appropriately based on the resolved formatter config.
   */
  hasValidFormattingForCharacterRanges(): boolean {
    return this.preciseFormatter.checkFormattingOfRanges(
      this.fileContents,
      this.formatterConfig,
      this.modifiedCharacterRanges,
    );
  }

  /**
   * Run the formatter on the file contents and store the result
   */
  formatCharacterRangesWithinContents(): void {
    this.formattedFileContents = this.preciseFormatter.formatRanges(
      this.fileContents,
      this.formatterConfig,
      this.modifiedCharacterRanges,
    );
  }

  /**
   * Return true if the formatted file contents are different to
   * what was originally resolved from disk.
   */
  shouldContentsBeUpdatedOnDisk(): boolean {
    return this.fileContents !== this.formattedFileContents;
  }

  /**
   * Write the updated file contents back to disk.
   */
  updateFileOnDisk(): void {
    writeFileSync(this.fullPath, this.formattedFileContents);
  }

  /**
   * We handle errors locally within this method to allow for
   * more granular feedback within the main() function of the
   * library.
   */
  calculateModifiedCharacterRanges(): { err: Error | null } {
    try {
      /**
       * Extract line change data from the git diff results.
       */
      const diff = getDiffForFile(
        this.gitDirectoryParent,
        this.fullPath,
        this.sha1,
        this.sha2,
      );
      const lineChangeData = extractLineChangeData(diff);
      /**
       * Convert the line change data into character data.
       */
      this.modifiedCharacterRanges = calculateCharacterRangesFromLineChanges(
        lineChangeData,
        this.fileContents,
      );
      return { err: null };
    } catch (err) {
      return { err };
    }
  }

  /**
   * Resolve and cache the relevant file contents.
   */
  private resolveFileContents() {
    this.fileContents = readFileSync(this.fullPath, 'utf8');
  }

  /**
   * Resolve and cache the relevant formatter config.
   */
  private resolveFormatterConfig() {
    this.formatterConfig = this.preciseFormatter.resolveConfig(this.fullPath);
  }
}
