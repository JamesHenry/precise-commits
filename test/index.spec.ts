import { readFileSync } from 'fs';

import { TestBed, readFixtures } from './test-utils';

import { main } from '../src/index';

const fixtures = readFixtures();
let testBed: TestBed;

describe('prettier-lines', () => {
  describe('main()', function() {
    beforeAll(() => {
      testBed = new TestBed();
    });

    fixtures.forEach(fixture => {
      it(fixture.fixtureName, () => {
        testBed.prepareFixtureInTmpDirectory(fixture);
        const tmpFile = testBed.getTmpFileForFixture(fixture);
        main(tmpFile.directoryPath, {
          checkOnly: false,
          filesWhitelist: null,
          sha1: null,
          sha2: null,
        });
        const formatted = readFileSync(tmpFile.path, 'utf8');
        expect(formatted).toMatchSnapshot();
      });
    });
  });

  describe('main() - checkOnly: true', function() {
    beforeAll(() => {
      testBed = new TestBed();
    });

    fixtures.forEach(fixture => {
      it(fixture.fixtureName, done => {
        expect.assertions(1);
        testBed.prepareFixtureInTmpDirectory(fixture);
        const tmpFile = testBed.getTmpFileForFixture(fixture);
        main(
          tmpFile.directoryPath,
          { checkOnly: true, filesWhitelist: null, sha1: null, sha2: null },
          {
            onInit() {},
            onBegunProcessingFile() {},
            onModifiedFilesDetected() {},
            onFinishedProcessingFile(_fn, _i, status) {
              expect(status).toEqual('INVALID_FORMATTING');
            },
            onComplete() {
              done();
            },
            onError(err) {
              done.fail(err);
            },
          },
        );
      });
    });
  });
});
