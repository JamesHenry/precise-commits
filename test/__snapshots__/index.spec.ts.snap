// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prettier-lines main() brand-new-jsx-file 1`] = `
"function HelloWorld({
  greeting = 'hello',
  greeted = '\\"World\\"',
  silent = false,
  onMouseOver,
}) {
  if (!greeting) {
    return null;
  }

  // TODO: Don't use random in render
  let num = Math.floor(Math.random() * 1e7)
    .toString()
    .replace(/\\\\.\\\\d+/gi, '');

  return (
    <div
      className=\\"HelloWorld\\"
      title={\`You are visitor number \${num}\`}
      onMouseOver={onMouseOver}
    >
      <strong>
        {greeting.slice(0, 1).toUpperCase() + greeting.slice(1).toLowerCase()}
      </strong>
      {greeting.endsWith(',') ? (
        ' '
      ) : (
        <span style={{ color: 'grey' }}>\\", \\"</span>
      )}
      <em>{greeted}</em>
      {silent ? '.' : '!'}
    </div>
  );
}
"
`;

exports[`prettier-lines main() empty-javascript-file 1`] = `
"var greeting = 'hello';
var audience = 'world';
"
`;

exports[`prettier-lines main() multi-range-regression 1`] = `
"import { execSync } from 'child_process';
import * as path from 'path';
import * as resolve from 'resolve';
import { getAffectedApps, getAppRoots, parseFiles } from './shared';

export function format(args: string[]) {
  const command = args[0];
  let patterns: string[];
  let rest: string[];

  try {
    if (args.length === 1) {
      patterns = ['\\"{apps,libs}/**/*.ts\\"'];
      rest = [];
    } else {
      const p = parseFiles(args.slice(1));
      patterns = p.files.filter(f => path.extname(f) === '.ts');
      rest = p.rest;

      const libsAndApp = rest.filter(a => a.startsWith('--libs-and-apps'))[0];
      if (libsAndApp) {
        patterns = getPatternsFromApps(patterns);
      }
    }
  } catch (e) {
    printError(command, e);
    process.exit(1);
  }

  switch (command) {
    case 'write':
      writePrettierLines(patterns);
      break;
    case 'check':
      checkPrettierLines(patterns);
      break;
  }
}

function getPatternsFromApps(affectedFiles: string[]): string[] {
  const roots = getAppRoots(getAffectedApps(affectedFiles));
  if (roots.length === 0) {
    return [];
  } else if (roots.length === 1) {
    return [\`\\\\\\"\${roots[0]}/**/*.ts\\\\\\"\`];
  } else {
    return [\`\\\\\\"{\${roots.join(',')}}/**/*.ts\\\\\\"\`];
  }
}

function printError(command: string, e: any) {
  console.error(\`Pass the SHA range, as follows: npm run format:\${command} -- SHA1 SHA2.\`);
  console.error(
    \`Or pass the list of files, as follows: npm run format:\${command} --files=\\"libs/mylib/index.ts,libs/mylib2/index.ts\\".\`
  );
  console.error(e.message);
}

function write(patterns: string[]) {
  if (patterns.length > 0) {
    execSync(\`node \${prettierPath()} --single-quote --print-width 120 --write \${patterns.join(' ')}\`, {
      stdio: [0, 1, 2]
    });
  }
}

function check(patterns: string[]) {
  if (patterns.length > 0) {
    try {
      execSync(\`node \${prettierPath()} --single-quote --print-width 120 --list-different \${patterns.join(' ')}\`, {
        stdio: [0, 1, 2]
      });
    } catch (e) {
      process.exit(1);
    }
  }
}

function prettierPath() {
  return \`\${path.dirname(resolve.sync('prettier', { basedir: __dirname }))}/bin-prettier.js\`;
}


function writePrettierLines(patterns: string[]) {
  if (patterns.length > 0) {
    let cmd = \`node \${prettierLinesPath()}\`;
    patterns.forEach(pattern => {
      cmd += \` --whitelist=\\"\${pattern}\\"\`;
    });
    execSync(cmd, {
      stdio: [0, 1, 2],
    });
  }
}

function checkPrettierLines(patterns: string[]) {
  if (patterns.length > 0) {
    try {
      let cmd = \`node \${prettierLinesPath()} --check-only\`;
      patterns.forEach(pattern => {
        cmd += \` --whitelist=\\"\${pattern}\\"\`;
      });
      execSync(cmd, {
        stdio: [0, 1, 2],
      });
    } catch (e) {
      process.exit(1);
    }
  }
}

function prettierLinesPath() {
  return \`\${path.dirname(
    resolve.sync('prettier-lines', { basedir: __dirname }),
  )}/bin/prettier-lines.js\`;
}
"
`;

exports[`prettier-lines main() multiple-updated-lines-javascript-file 1`] = `
"var     untouchedUglyLines1 = ''
function TouchedFunctionSignature({
  greeting = \\"Hello\\",
  greeted = '\\"Mars\\"',
  silent = false
}) {}
const newlyAddedUglyLine = true;"
`;

exports[`prettier-lines main() single-updated-line-typescript 1`] = `
"const a = true
const b = true;"
`;
